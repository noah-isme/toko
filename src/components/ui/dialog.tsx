"use client";

import * as DialogPrimitive from "@radix-ui/react-dialog";
import { X } from "lucide-react";
import * as React from "react";

import { cn } from "@/lib/utils";

const Dialog = DialogPrimitive.Root;
const DialogTrigger = DialogPrimitive.Trigger;
const DialogPortal = DialogPrimitive.Portal;
const DialogClose = DialogPrimitive.Close;
const DialogTitle = DialogPrimitive.Title;
const DialogDescription = DialogPrimitive.Description;

const DialogOverlay = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Overlay>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Overlay>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Overlay
    ref={ref}
    className={cn(
      "fixed inset-0 z-50 bg-black/40 backdrop-blur-sm data-[state=open]:animate-in data-[state=closed]:animate-out",
      className,
    )}
    {...props}
  />
));
DialogOverlay.displayName = DialogPrimitive.Overlay.displayName;

const DialogContent = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content>
>(({ className, children, ...props }, ref) => (
  <DialogPortal>
    <DialogOverlay />
    <DialogPrimitive.Content
      ref={ref}
      className={cn(
        "fixed inset-y-0 right-0 z-50 flex w-full max-w-lg flex-col bg-background shadow-lg transition-all",
        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:slide-out-to-right",
        "data-[state=open]:slide-in-from-right",
        className,
      )}
      {...props}
    >
      {children}
      <DialogPrimitive.Close className="absolute right-4 top-4 rounded-sm opacity-70 transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none">
        <X className="h-4 w-4" />
        <span className="sr-only">Close</span>
      </DialogPrimitive.Close>
    </DialogPrimitive.Content>
  </DialogPortal>
));
DialogContent.displayName = DialogPrimitive.Content.displayName;

const DialogHeader = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div className={cn("flex flex-col space-y-1.5 p-6", className)} {...props} />
);
DialogHeader.displayName = "DialogHeader";

const DialogFooter = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    className={cn("flex flex-col-reverse gap-2 p-6 sm:flex-row sm:justify-end", className)}
    {...props}
  />
);
DialogFooter.displayName = "DialogFooter";

export {
  Dialog,
  DialogPortal,
  DialogOverlay,
  DialogTrigger,
  DialogContent,
  DialogHeader,
  DialogFooter,
  DialogTitle,
  DialogDescription,
  DialogClose,
};
