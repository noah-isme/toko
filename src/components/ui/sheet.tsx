'use client';

import * as DialogPrimitive from '@radix-ui/react-dialog';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '@/lib/utils';

const Sheet = DialogPrimitive.Root;
const SheetTrigger = DialogPrimitive.Trigger;
const SheetClose = DialogPrimitive.Close;

const SheetPortal = DialogPrimitive.Portal;

const SheetOverlay = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Overlay>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Overlay>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Overlay
    ref={ref}
    className={cn('fixed inset-0 z-50 bg-black/80', className)}
    {...props}
  />
));
SheetOverlay.displayName = DialogPrimitive.Overlay.displayName;

const sheetVariants = cva(
  'fixed z-50 gap-4 bg-background p-6 shadow-lg transition ease-in-out data-[state=open]:animate-in data-[state=closed]:animate-out',
  {
    variants: {
      side: {
        top: 'inset-x-0 top-0 border-b data-[state=closed]:slide-out-to-top data-[state=open]:slide-in-from-top',
        bottom:
          'inset-x-0 bottom-0 border-t data-[state=closed]:slide-out-to-bottom data-[state=open]:slide-in-from-bottom',
        left: 'inset-y-0 left-0 h-full w-3/4 border-r data-[state=closed]:slide-out-to-left data-[state=open]:slide-in-from-left sm:max-w-sm',
        right:
          'inset-y-0 right-0 h-full w-3/4 border-l data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-sm',
      },
    },
    defaultVariants: {
      side: 'right',
    },
  },
);

const SheetContent = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content> &
    VariantProps<typeof sheetVariants>
>(({ side = 'right', className, children, ...props }, ref) => (
  <SheetPortal>
    <SheetOverlay />
    <DialogPrimitive.Content
      ref={ref}
      className={cn(sheetVariants({ side }), className)}
      {...props}
    >
      {children}
    </DialogPrimitive.Content>
  </SheetPortal>
));
SheetContent.displayName = DialogPrimitive.Content.displayName;

const SheetHeader = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div className={cn('grid gap-1.5 text-left', className)} {...props} />
);
SheetHeader.displayName = 'SheetHeader';

const SheetFooter = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div className={cn('mt-auto flex flex-col gap-2', className)} {...props} />
);
SheetFooter.displayName = 'SheetFooter';

const SheetTitle = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Title>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Title>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Title ref={ref} className={cn('text-lg font-semibold', className)} {...props} />
));
SheetTitle.displayName = DialogPrimitive.Title.displayName;

const SheetDescription = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Description>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Description>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Description
    ref={ref}
    className={cn('text-sm text-muted-foreground', className)}
    {...props}
  />
));
SheetDescription.displayName = DialogPrimitive.Description.displayName;

export {
  Sheet,
  SheetTrigger,
  SheetClose,
  SheetContent,
  SheetHeader,
  SheetFooter,
  SheetTitle,
  SheetDescription,
};
