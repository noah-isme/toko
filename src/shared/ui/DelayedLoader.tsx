'use client';

import { Loader2 } from 'lucide-react';
import { useEffect, useRef, useState } from 'react';

import { cn } from '@/lib/utils';

export type DelayedLoaderVariant = 'inline' | 'overlay';

export interface DelayedLoaderProps {
  active: boolean;
  variant?: DelayedLoaderVariant;
  label?: string;
  delayMs?: number;
  className?: string;
}

const DEFAULT_DELAY = 400;

export function DelayedLoader({
  active,
  variant = 'inline',
  label = 'Memuatâ€¦',
  delayMs = DEFAULT_DELAY,
  className,
}: DelayedLoaderProps) {
  const timerRef = useRef<ReturnType<typeof setTimeout>>();
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    if (timerRef.current) {
      clearTimeout(timerRef.current);
      timerRef.current = undefined;
    }

    if (!active) {
      setVisible(false);
      return;
    }

    timerRef.current = setTimeout(() => {
      setVisible(true);
    }, delayMs);

    return () => {
      if (timerRef.current) {
        clearTimeout(timerRef.current);
      }
    };
  }, [active, delayMs]);

  useEffect(() => {
    if (!active) {
      setVisible(false);
    }
  }, [active]);

  if (!visible) {
    return null;
  }

  if (variant === 'overlay') {
    return (
      <div
        aria-live="polite"
        role="status"
        className={cn(
          'absolute inset-0 z-20 pointer-events-none flex items-center justify-center rounded-lg bg-background/70 backdrop-blur-sm',
          className,
        )}
      >
        <Spinner label={label} />
      </div>
    );
  }

  return (
    <span
      aria-live="polite"
      role="status"
      className={cn('inline-flex items-center gap-2 text-sm text-muted-foreground', className)}
    >
      <Spinner label={label} />
    </span>
  );
}

function Spinner({ label }: { label: string }) {
  return (
    <span className="inline-flex items-center gap-2">
      <Loader2 className="prm:no-anim h-4 w-4 animate-spin" aria-hidden="true" />
      <span>{label}</span>
    </span>
  );
}
