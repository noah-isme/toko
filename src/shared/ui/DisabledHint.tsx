'use client';

import React, { useId } from 'react';

import { cn } from '@/lib/utils';

export interface DisabledHintProps {
  id?: string;
  message: string;
  className?: string;
  tone?: 'info' | 'warning';
}

export function DisabledHint({
  id: providedId,
  message,
  className,
  tone = 'info',
}: DisabledHintProps) {
  const autoId = useId();
  const id = providedId ?? autoId;

  return (
    <p
      id={id}
      role="note"
      className={cn(
        'mt-2 text-xs leading-relaxed',
        tone === 'warning' ? 'text-amber-600 dark:text-amber-400' : 'text-muted-foreground',
        className,
      )}
    >
      {message}
    </p>
  );
}
