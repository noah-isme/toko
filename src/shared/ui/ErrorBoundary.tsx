'use client';

import React from 'react';

import { normalizeError } from '@/shared/lib/normalizeError';

type ErrorBoundaryProps = {
  fallback?: React.ReactNode;
  children?: React.ReactNode;
};

type ErrorBoundaryState = {
  hasError: boolean;
  message?: string;
};

const DEFAULT_ERROR_MESSAGE = 'Terjadi kesalahan. Coba lagi.';

export class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  state: ErrorBoundaryState = {
    hasError: false,
    message: undefined,
  };

  static getDerivedStateFromError(): ErrorBoundaryState {
    return { hasError: true };
  }

  componentDidCatch(error: unknown, errorInfo: React.ErrorInfo): void {
    if (process.env.NODE_ENV !== 'production') {
      console.error('ErrorBoundary caught an error', error, errorInfo);
    }

    let message = DEFAULT_ERROR_MESSAGE;

    try {
      message = normalizeError(error);
    } catch (normalizeErrorFailure) {
      if (process.env.NODE_ENV !== 'production') {
        console.error('Failed to normalize error', normalizeErrorFailure);
      }
    }

    this.setState({ hasError: true, message });
  }

  private handleReload = () => {
    if (typeof window !== 'undefined' && window.location) {
      window.location.reload();
    }
  };

  render() {
    const { hasError, message } = this.state;
    const { fallback, children } = this.props;

    if (hasError) {
      if (fallback) {
        return fallback;
      }

      return (
        <div
          role="alert"
          className="flex flex-col items-start gap-3 rounded-lg border border-destructive/40 bg-destructive/10 p-6 text-sm text-foreground"
        >
          <div className="space-y-1">
            <h2 className="text-base font-semibold text-destructive">Terjadi kesalahan</h2>
            <p className="text-muted-foreground">{message ?? DEFAULT_ERROR_MESSAGE}</p>
          </div>
          <button
            type="button"
            onClick={this.handleReload}
            className="inline-flex items-center justify-center rounded-md bg-destructive px-4 py-2 text-sm font-medium text-destructive-foreground shadow transition-colors hover:bg-destructive/90 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-destructive focus-visible:ring-offset-2"
          >
            Muat Ulang
          </button>
        </div>
      );
    }

    return (children as React.ReactNode) ?? null;
  }
}

export default ErrorBoundary;
