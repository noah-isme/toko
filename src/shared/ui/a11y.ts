import { createElement } from 'react';
import type { ReactElement } from 'react';

export function srOnly(text: string): ReactElement {
  return createElement('span', { className: 'sr-only' }, text);
}
