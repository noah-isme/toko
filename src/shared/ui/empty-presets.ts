import type { EmptyStateProps } from './EmptyState';

export function emptyProducts(): EmptyStateProps {
  return {
    title: 'Produk belum tersedia',
    description:
      'Kami tidak menemukan produk yang sesuai. Coba ubah filter atau periksa kembali nanti.',
    cta: {
      label: 'Lihat semua produk',
      href: '/products',
    },
  };
}

export function emptyCart(): EmptyStateProps {
  return {
    title: 'Keranjang belanja kosong',
    description: 'Mulai tambahkan produk ke keranjang Anda untuk melanjutkan ke proses checkout.',
    cta: {
      label: 'Belanja sekarang',
      href: '/products',
    },
  };
}

export function emptyOrders(): EmptyStateProps {
  return {
    title: 'Belum ada pesanan',
    description: 'Riwayat pesanan Anda akan muncul di sini setelah berhasil melakukan checkout.',
    cta: {
      label: 'Lanjutkan belanja',
      href: '/products',
    },
  };
}
