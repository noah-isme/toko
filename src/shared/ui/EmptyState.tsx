import Link from 'next/link';
import * as React from 'react';
import type { ComponentProps, ReactNode } from 'react';

import { Button } from '@/components/ui/button';
import { cn } from '@/lib/utils';

export interface EmptyStateProps {
  icon?: ReactNode;
  title: string;
  description?: string;
  cta?: {
    label: string;
    href: ComponentProps<typeof Link>['href'];
  };
  className?: string;
}

export function EmptyState({ icon, title, description, cta, className }: EmptyStateProps) {
  return (
    <React.Fragment>
      <div
        className={cn(
          'flex flex-col items-center justify-center gap-4 rounded-lg border border-dashed border-muted-foreground/30 bg-muted/10 px-6 py-16 text-center',
          className,
        )}
      >
        {icon ? (
          <div className="text-muted-foreground [&>svg]:h-12 [&>svg]:w-12">{icon}</div>
        ) : null}
        <div className="space-y-2">
          <h2 className="text-xl font-semibold text-foreground">{title}</h2>
          {description ? (
            <p className="mx-auto max-w-md text-sm text-muted-foreground">{description}</p>
          ) : null}
        </div>
        {cta ? (
          <Button asChild>
            <Link href={cta.href}>{cta.label}</Link>
          </Button>
        ) : null}
      </div>
    </React.Fragment>
  );
}
