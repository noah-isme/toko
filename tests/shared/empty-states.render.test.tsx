import { cleanup, render, screen } from '@testing-library/react';
import React from 'react';
import { afterEach, describe, expect, it } from 'vitest';

import { emptyCart, emptyOrders, emptyProducts } from '@/shared/ui/empty-presets';
import { EmptyState } from '@/shared/ui/EmptyState';

describe('EmptyState component', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders title without optional props', () => {
    const { container } = render(<EmptyState title="Tidak ada data" />);

    expect(screen.getByText('Tidak ada data')).toBeInTheDocument();
    expect(screen.queryByRole('link')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders description and CTA when provided', () => {
    render(
      <EmptyState
        title="Keranjang kosong"
        description="Tambahkan produk untuk melanjutkan."
        cta={{ label: 'Belanja', href: '/products' }}
      />,
    );

    expect(screen.getByText('Tambahkan produk untuk melanjutkan.')).toBeInTheDocument();
    const ctaLink = screen.getByRole('link', { name: 'Belanja' });
    expect(ctaLink).toBeInTheDocument();
    expect(ctaLink).toHaveAttribute('href', '/products');
  });
});

describe('Empty state presets', () => {
  const presets = [
    ['emptyProducts', emptyProducts],
    ['emptyCart', emptyCart],
    ['emptyOrders', emptyOrders],
  ] as const;

  it.each(presets)('%s provides title, description, and CTA', (_, preset) => {
    const props = preset();

    expect(props.title).toBeTruthy();
    expect(props.description).toBeTruthy();
    expect(props.cta?.label).toBeTruthy();
    expect(props.cta?.href).toBeTruthy();
  });
});
