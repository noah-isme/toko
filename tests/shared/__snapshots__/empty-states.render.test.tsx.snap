// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmptyState component > renders title without optional props 1`] = `
<div
  class="flex flex-col items-center justify-center gap-4 rounded-lg border border-dashed border-muted-foreground/30 bg-muted/10 px-6 py-16 text-center"
>
  <div
    class="space-y-2"
  >
    <h2
      class="text-xl font-semibold text-foreground"
    >
      Tidak ada data
    </h2>
  </div>
</div>
`;
